mmrazor/models/task_modules/estimators/counters/flops_params_counter.py

```diff
@@ -0,0 +1,467 @@
+    Args:
+        model (nn.Module): The model for complexity calculation.
+        input_shape (tuple): Input shape (including batchsize) used for
+            calculation.
+        spec_modules (list): A list that contains the names of several spec
+            modules, which users want to get resources infos of them.
+            e.g., ['backbone', 'head'], ['backbone.layer1']. Default to [].
+        print_per_layer_stat (bool): Whether to print complexity information
+            for each layer in a model. Default to True.
+        as_strings (bool): Output FLOPs and params counts in a string form.
+            Default to True.
+        input_constructor (None | callable): If specified, it takes a callable
+            method that generates input. otherwise, it will generate a random
+            tensor with input shape to calculate FLOPs. Default to None.
+        flush (bool): same as that in :func:`print`. Default to False.
+        ost (stream): same as ``file`` param in :func:`print`.
+            Default to sys.stdout.
+        disabled_counters (list): One can limit which ops' spec would be
+            calculated. Default to `None`.
+
+    Returns:
+        tuple[float | str] | dict[str, float]: If `as_strings` is set to True,
+            it will return FLOPs and parameter counts in a string format.
+            Otherwise, it will return those in a float number format.
+            If len(spec_modules) > 0, it will return a resource info dict with
+            FLOPs and parameter counts of each spec module in float format.
+    """
+    assert type(input_shape) is tuple
+    assert len(input_shape) >= 1
+    assert isinstance(model, nn.Module)
+    flops_params_model = add_flops_params_counting_methods(model)
+    flops_params_model.eval()
+    flops_params_model.start_flops_params_count(disabled_counters)
+    if input_constructor:
+        input = input_constructor(input_shape)
+        _ = flops_params_model(**input)
+    else:
+        try:
+            batch = torch.ones(()).new_empty(
+                tuple(input_shape),
+                dtype=next(flops_params_model.parameters()).dtype,
+                device=next(flops_params_model.parameters()).device)
+        except StopIteration:
+            # Avoid StopIteration for models which have no parameters,
+            # like `nn.Relu()`, `nn.AvgPool2d`, etc.
+            batch = torch.ones(()).new_empty(tuple(input_shape))
+
+        _ = flops_params_model(batch)
+
+    flops_count, params_count = \
+        flops_params_model.compute_average_flops_params_cost()
+
+    if print_per_layer_stat:
+        print_model_with_flops_params(
+            flops_params_model,
+            flops_count,
+            params_count,
+            ost=ost,
+            flush=flush)
+
+    if len(spec_modules):
+        spec_modules_resources = dict()
+        accumulate_sub_module_flops_params(flops_params_model)
+        for name, module in flops_params_model.architecture.named_modules():
```

Comment:
Not all the flops_params_model have the `architecture` attribute.

Reply:
Done.